pkg/operation/botanist/dns.go

```diff
@@ -225,6 +270,36 @@ func (b *Botanist) GenerateTerraformDNSConfig(name, hostedZoneID, targetType str
 	}
 }
 
+// generateTerraformAlicloudDNSConfig To adapt Alicloud Terraform,
+// @input Param domain should be split as name and host_record.
+// name is registered in Alicloud. It is always like xxx.xxx. If there is an exception, this function should be rewritten !!!
+// host_record is the host name in A or CNAME record
+func (b *Botanist) generateTerraformAlicloudDNSConfig(domain, targetType string, value string, purposeInternalDomain bool) (map[string]interface{}, error) {
+	secret, err := b.getDomainCredentials(purposeInternalDomain)
+	if err != nil {
+		return nil, err
+	}
+	name, ok := secret.Annotations[common.DNSDomain]
+	// if the secret is not default domain secret or internal domain secret
+	if !ok {
+		// domain should be generated from shoot.spec.dns.domain
+		splits := strings.Split(domain, ".")
+		// Shoot validation promises the shoot.spec.dns.domain has at least one ".", and domain has more than one "."
+		l := len(splits)
+		name = fmt.Sprintf("%s.%s", splits[l-2], splits[l-1])
```

Comment:
I understood that the `hostRecord` has to be the part of the DNS record **without** the domain, i.e. if the DNS record is `api.shoot.example.com` the `hostRecord` would be `api.shoot`, right?

Here you are cutting the last two parts, i.e. in the above example `example.com`. What happens if the domain is longer, i.e. `my.first.landscape.example.com` with a shoot domain `my-shoot.my-project.my.first.landscape.example.com`? The `hostRecord` would be `my-shoot.my-project.my.first.landscape`, does this still work/is this behaviour fine? Or should it rather be `my-shoot.my-project`?

Reply:
@rfranzke  Good question.
 
 > I understood that the hostRecord has to be the part of the DNS record without the domain, i.e. if the DNS record is api.shoot.example.com the hostRecord would be api.shoot, right?

You are right. And I tested OK.

In the Second part, from my experience of usage of Alicloud, only top level domain (in xxx.xxx format) could be the name, because alicloud only **owns** top level domain.
In your case though domain my.first.landscape.example.com is specified in shoot DNS, it means the tenant owns the domain **example.com** in alicloud. If the generated domain is my-shoot.my-project.my.first.landscape.example.com, we still need to split into 2 parts: my-shoot.my-project.my.first.landscape and example.com.

Usually, the tenant probably has a second shoot with domain set my.second.landscape.example.com. Generated domain will be my-shoot.my-project.my.second.landscape.example.com and my code will handle the splitting logic.

In summary, we CAN'T use shoot.spec.dns.domain as name in terraform DNS record source.

Did I make it clear?

Reply:
Interesting behaviour. Anyway, you are the expert, I have no experience with Alicloud. I just wanted to understand this part. So, if you say it behaves correctly then let's keep it like this. It does, right?